### PR TITLE
Feature/topics endpoint

### PR DESCRIPTION
**Pull request by SOY4RIAS**
### Pull request type

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

---

### Description

- Topics CRUD endpoints added

---

#### Board reference

- [board](link)

### CODE REVIEW

File: src/api/topics/topic.entity.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import {
+  Column,
+  CreateDateColumn,
+  Entity,
+  PrimaryGeneratedColumn,
+} from 'typeorm';
+
+@Entity('topics')
+export class TopicEntity {
+  @PrimaryGeneratedColumn()
```
Comment (brunoBrizo) [inline src/api/topics/topic.entity.ts]: WDYT about creating like a "base" entity with common attributes? Something like id, create at, updated at. And then you can just extend it with all other entities.

File: src/api/topics/topics.service.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  findOne(id: number): Promise<TopicEntity | null> {
+    return this.topicRepository.findOne({
+      where: { id },
+    });
+  }
+
+  update(id: number, updateTopicDto: UpdateTopicDto): Promise<UpdateResult> {
+    const topic = this.findOne(id);
```
Comment (brunoBrizo) [inline src/api/topics/topics.service.ts]: The method findOne returns a promise, so here you should await it

**Review by brunoBrizo — CHANGES_REQUESTED** [review]
Great job! Just left a couple of comments to be reviewed.

File: src/api/topics/topics.service.ts (hunk 2 above)
Comment (SOY4RIAS) [reply src/api/topics/topics.service.ts]: 😅 yup, I think I accidentally deleted it

File: src/api/topics/topics.controller.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import { UpdateTopicDto } from '@api/topics/dto/update-topic.dto';
+import { TopicEntity } from '@api/topics/topic.entity';
+import { TopicsService } from '@api/topics/topics.service';
+
+@Controller('topics')
+export class TopicsController {
+  constructor(private readonly topicsService: TopicsService) {}
+
+  @Post()
+  create(@Body() createTopicDto: CreateTopicDto) {
```
Comment (mds2000) [inline src/api/topics/topics.controller.ts:21]: All methods returning a Promise should be async, this ensures the proper functioning of the Node.js event loop.
Also, they should always have a return type

File: src/api/topics/topics.controller.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    return this.topicsService.create(createTopicDto);
+  }
+
+  @Get()
+  findAll(): Promise<TopicEntity[]> {
+    return this.topicsService.findAll();
+  }
+
+  @Get(':id')
+  findOne(@Param('id') id: string): Promise<TopicEntity | null> {
```
Comment (mds2000) [inline src/api/topics/topics.controller.ts:31]: Here you could use the built-in nest pipe to ensure the param id is an integer
```suggestion
  findOne(@Param('id', ParseIntPipe) id: number): Promise<TopicEntity | null> {
```

**Review by mds2000 — APPROVED** [review]
Great work! Just left some comments. Are you planning in doing the tests in another pr?

**Review by brunoBrizo — APPROVED** [review]
Great work! LGTM

File: src/api/topics/topics.service.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  create(createTopicDto: CreateTopicDto): Promise<TopicEntity> {
+    return this.topicRepository.save(createTopicDto);
+  }
+
+  findAll(): Promise<TopicEntity[]> {
+    return this.topicRepository.find();
+  }
+
+  findOne(id: number): Promise<TopicEntity | null> {
+    return this.topicRepository.findOne({
```
Comment (matiasirpac) [inline src/api/topics/topics.service.ts:25]: Here, you could use "findOneOrFail" and then handle the error. So, when you reuse this method, the error is always caught. For example:

  ```
async findById(id: number): Promise<TopicEntity> {
    try {
      const topic = await this.topicRepository.findOneOrFail({ id });
      return topic;
    } catch (error) {
      throw new TopicNotFoundException(id);
    }
  }
```
  
  Then in update:
  
  ```
async update(
    id: number,
    updateTopicDto: UpdateTopicDto,
  ): Promise<UpdateResult> {
    const topic = await this.findById(id);
    try {
      return await this.topicRepository.update({ id: topic.id }, updateTopicDto);
      return topic;
    } catch (error) {
      throw new TopicNotUpdateException(id, error);
    }
  }
```
  
  This way each method handles its errors and the code is cleaner, I even suggest changing the name of findOne to findById to make it more descriptive.

What do you think?

**Review by matiasirpac — APPROVED** [review]
Good job!

**Pull request merged**